sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala

```diff
@@ -423,4 +429,20 @@ object CTESubstitution extends Rule[LogicalPlan] {
       case _ => WithCTE(p, cteDefs)
     }
   }
+
+  /**
+   * Counts number of self-references in a recursive CTE definition and throws an error
+   * if that number is bigger than 1.
+   */
+  private def checkNumberOfSelfReferences(cteDef: CTERelationDef): Unit = {
+    val numOfSelfRef = cteDef.map[Boolean] {
+      case CTERelationRef(cteDef.id, _, _, _, _, true) => true
+      case other => false
+    }.count(_ == true)
```

Comment:
I think you can drop the `(_ == true)` or the `== true` and it will have the same result with simpler code.

Reply:
or use .collect
```
val numOfSelfRef = cteDef.collectWithSubqueries {
  case ref: CTERelationRef red.cteId = cteDef.id => ref
}.length
```